examples/pytorch/language-modeling/run_fim.py

```diff
@@ -0,0 +1,799 @@
+        "token": model_args.token,
+        "trust_remote_code": model_args.trust_remote_code,
+    }
+    if model_args.config_name:
+        config = AutoConfig.from_pretrained(model_args.config_name, **config_kwargs)
+    elif model_args.model_name_or_path:
+        config = AutoConfig.from_pretrained(model_args.model_name_or_path, **config_kwargs)
+    else:
+        config = CONFIG_MAPPING[model_args.model_type]()
+        logger.warning("You are instantiating a new config instance from scratch.")
+        if model_args.config_overrides is not None:
+            logger.info(f"Overriding config: {model_args.config_overrides}")
+            config.update_from_string(model_args.config_overrides)
+            logger.info(f"New config: {config}")
+
+    tokenizer_kwargs = {
+        "cache_dir": model_args.cache_dir,
+        "use_fast": model_args.use_fast_tokenizer,
+        "revision": model_args.model_revision,
+        "token": model_args.token,
+        "trust_remote_code": model_args.trust_remote_code,
+    }
+    if model_args.tokenizer_name:
+        tokenizer = AutoTokenizer.from_pretrained(model_args.tokenizer_name, **tokenizer_kwargs)
+    elif model_args.model_name_or_path:
+        tokenizer = AutoTokenizer.from_pretrained(model_args.model_name_or_path, **tokenizer_kwargs)
+    else:
+        raise ValueError(
+            "You are instantiating a new tokenizer from scratch. This is not supported by this script. "
+            "You can do it from another script, save it, and load it from here, using --tokenizer_name."
+        )
+
+    if model_args.model_name_or_path:
+        torch_dtype = (
+            model_args.torch_dtype
+            if model_args.torch_dtype in ["auto", None]
+            else getattr(torch, model_args.torch_dtype)
+        )
+        model = AutoModelForCausalLM.from_pretrained(
+            model_args.model_name_or_path,
+            from_tf=bool(".ckpt" in model_args.model_name_or_path),
+            config=config,
+            cache_dir=model_args.cache_dir,
+            revision=model_args.model_revision,
+            token=model_args.token,
+            trust_remote_code=model_args.trust_remote_code,
+            torch_dtype=torch_dtype,
+            low_cpu_mem_usage=model_args.low_cpu_mem_usage,
+        )
+
+    else:
+        model = AutoModelForCausalLM.from_config(config, trust_remote_code=model_args.trust_remote_code)
+        n_params = sum({p.data_ptr(): p.numel() for p in model.parameters()}.values())
+        logger.info(f"Training new model from scratch - Total size={n_params/2**20:.2f}M params")
+
+    # Add the new FIM tokens to the tokenizer and resize model's vocab embeddings
+    special_tokens = [data_args.fim_prefix_token, data_args.fim_middle_token, data_args.fim_suffix_token]
+    if data_args.truncate_or_pad:
+        special_tokens.append(data_args.fim_pad_token)
+
+    tokenizer.add_tokens(special_tokens)
+    model.resize_token_embeddings(len(tokenizer))
```

Comment:
resizing should be handled like this: https://nlp.stanford.edu/~johnhew/vocab-expansion.html

Reply:
No what I mean is using the average of the already available tokens not necessarly wte. You can always get the embeding with `get_input_embeddings` and `set_input_embeddings`

Reply:
@ArthurZucker I am doing this for adding new tokens' embeddings: they are set to the average of the existing embeddings. Is this a good approach?

```python
model.resize_token_embeddings(len(model.get_input_embeddings().weight) + len(new_tokens))

# Get the updated embedding layer
embeddings = model.get_input_embeddings()

# Calculate the average of the existing embeddings
avg_embedding = embeddings.weight[:-len(new_tokens)].mean(dim=0)
new_embeddings = embeddings.weight.clone()

# Set the new tokens' embeddings to this average
new_embeddings[-len(new_tokens):] = avg_embedding

# Update the model's embedding layer
embeddings.weight = torch.nn.Parameter(new_embeddings)
model.set_input_embeddings(embeddings)
```

Reply:
It is, but this will make the embeddings equal, you need to sample from a uniform distribution centered around the mean of the embeddings! But all good otherwise

Reply:
Thanks, @ArthurZucker , I have a doubt,
When using `pad_to_mulitple_of` argument, should the tokens `len(new_tokens)` actually reflect all the new tokens added, including the ones added due to padding?

For example, if `4` FIM tokens and extra `3` tokens were added to GPT2 to make the embedding size `50264` (multiple of 8), should the embeddings' sampling be done all 7 tokens or just the original `4` FIM tokens only?

Reply:
I think the documentation should explain this but the idea is you just say you want len(tokenizer), the pad_to_multiple of has nothing to do with this. You should set it according to the device you are using

Reply:
I see, thanks for clarifying this!